src/error.rs

```diff
@@ -0,0 +1,185 @@
+                };
+                write!(f, "sheet {message} not found")
+            }
+            FastExcelErrorKind::ArrowError(err) => write!(f, "arrow error: {err}"),
+            FastExcelErrorKind::InvalidParameters(err) => write!(f, "invalid parameters: {err}"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct FastExcelError {
+    kind: FastExcelErrorKind,
+    context: Vec<String>,
+}
+
+pub(crate) trait ErrorContext {
+    fn with_context<S: ToString, F>(self, ctx_fn: F) -> Self
+    where
+        F: FnOnce() -> S;
+}
+
+impl FastExcelError {
+    pub(crate) fn new(kind: FastExcelErrorKind) -> Self {
+        Self {
+            kind,
+            context: vec![],
+        }
+    }
+}
+
+impl Display for FastExcelError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{kind}", kind = self.kind)?;
+        if !self.context.is_empty() {
+            writeln!(f, "\nContext:")?;
+
+            self.context
+                .iter()
+                .enumerate()
+                .try_for_each(|(idx, ctx_value)| writeln!(f, "    {idx}: {ctx_value}"))?;
+        }
+        Ok(())
+    }
+}
+
+impl Error for FastExcelError {}
+
+impl ErrorContext for FastExcelError {
+    fn with_context<S: ToString, F>(mut self, ctx_fn: F) -> Self
+    where
+        F: FnOnce() -> S,
+    {
+        self.context.push(ctx_fn().to_string());
+        self
+    }
+}
+
+impl From<FastExcelErrorKind> for FastExcelError {
+    fn from(kind: FastExcelErrorKind) -> Self {
+        FastExcelError::new(kind)
+    }
+}
+
+pub(crate) type FastExcelResult<T> = Result<T, FastExcelError>;
+
+impl<T> ErrorContext for FastExcelResult<T> {
+    fn with_context<S: ToString, F>(self, ctx_fn: F) -> Self
+    where
+        F: FnOnce() -> S,
+    {
+        match self {
+            Ok(_) => self,
+            Err(e) => Err(e.with_context(ctx_fn)),
+        }
+    }
+}
+
+/// Contains Python versions of our custom errors
+pub(crate) mod py_errors {
+    use super::FastExcelErrorKind;
+    use pyo3::{create_exception, exceptions::PyException, PyResult};
+
+    // Base fastexcel error
+    create_exception!(_fastexcel, FastExcelError, PyException);
+    // Unsupported column type
+    create_exception!(
+        _fastexcel,
+        UnsupportedColumnTypeCombinationError,
+        FastExcelError
+    );
+    // Cannot retrieve cell data
+    create_exception!(_fastexcel, CannotRetrieveCellDataError, FastExcelError);
+    // Calamine cell error
+    create_exception!(_fastexcel, CalamineCellError, FastExcelError);
+    // Calamine error
+    create_exception!(_fastexcel, CalamineError, FastExcelError);
+    // Sheet not found
+    create_exception!(_fastexcel, SheetNotFoundError, FastExcelError);
+    // Arrow error
+    create_exception!(_fastexcel, ArrowError, FastExcelError);
+    // Invalid parameters
+    create_exception!(_fastexcel, InvalidParametersError, FastExcelError);
+
+    pub(crate) trait IntoPyResult {
```

Comment:
TryInto since it returns a result?

Reply:
Yes but the transformation into a `PyResult` cannot fail. To me, `TryIntoPyResult` would imply

```
fn try_into_pyresult(self) -> Result<PyResult<Inner>, _>;
```

Reply:
Ahh yes yes ofc